google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -43,35 +43,93 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
 
 namespace {
 
-opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeParentSpan(
+/// Inserts a link for each span in @p destination between the @p begin and @p
+/// end iterators.
+template <typename Iterator, typename OutputIterator>
+void GenerateLinks(Iterator begin, Iterator end, OutputIterator destination) {
+  static_assert(
+      std::is_same<
+          absl::decay_t<decltype(*begin)>,
+          opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>::value,
+      "Iterator is not the right type.");
+  using AttributesList =
+      std::vector<std::pair<opentelemetry::nostd::string_view,
+                            opentelemetry::common::AttributeValue>>;
+  std::transform(begin, end, destination,
+                 [i = static_cast<std::int64_t>(0)](auto const& span) mutable {
+                   return std::make_pair(
+                       span->GetContext(),
+                       AttributesList{{"messaging.pubsub.message.link", i++}});
+                 });
+}
+
+template <typename T>
+void GenerateBatchSinkChildrenSpans(std::vector<T> const& message_spans,
+                                    T batch_sink_parent_span,
+                                    std::ptrdiff_t batch_size,
+                                    std::vector<T>& batch_sink_spans) {
+  using LinksList = std::vector<
+      std::pair<opentelemetry::trace::SpanContext,
+                std::vector<std::pair<opentelemetry::nostd::string_view,
+                                      opentelemetry::common::AttributeValue>>>>;
+  auto cut = [&](auto i) {
+    return std::next(
+        i, std::min(batch_size - 1, std::distance(i, message_spans.end())));
+  };
+  int count = 0;
+  for (auto i = message_spans.begin(); i != message_spans.end(); i = cut(i)) {
+    LinksList links;
+    // Generate links between [i, min((i + batch_size) -1), end)) range.
```

Comment:
I think you really want $[i, min(i + batchSize, end))$.  It is an semi-open range, when there are exactly 128 elements you want $[0, 128)$. You may be thinking $[0, 127]$.

Reply:
Thanks, that is what I meant. Fixed the comment